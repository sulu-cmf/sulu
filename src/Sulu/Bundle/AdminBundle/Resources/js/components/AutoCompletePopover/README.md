This component is the popover showing the suggestions in the [`SingleAutoComplete`](#singleautocomplete) and
[`MultiAutoComplete`](#multiautocomplete) field.
