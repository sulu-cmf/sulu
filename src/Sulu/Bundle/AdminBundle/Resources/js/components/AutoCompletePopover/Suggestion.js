// @flow
import React from 'react';
import type {Node} from 'react';
import Icon from '../Icon';
import suggestionStyles from './suggestion.scss';

type Props = {|
    children: string | (highlight: (text: string) => Node) => Node,
    icon?: string,
    minWidth: number,
    onSelect: (value: Object) => void,
    query: ?string,
    value: Object,
|};

export default class Suggestion extends React.PureComponent<Props> {
    static defaultProps = {
        minWidth: 0,
        query: '',
    };

    highlightMatchingTextPart = (text: string) => {
        if (!text) {
            return;
        }

        const query = this.props.query || '';
        const regex = new RegExp(query, 'gi');
        const matches = text.match(regex);

        if (!matches || query.length === 0) {
            return text;
        }

        let matchIndex = 0;
        const highlightedMatches = text.replace(regex, () => {
            return `<strong>${matches[matchIndex++]}</strong>`;
        });

        return (
            <span dangerouslySetInnerHTML={{ __html: highlightedMatches }} />
        );
    };

    handleClick = () => {
        const {
            value,
            onSelect,
        } = this.props;

        if (onSelect) {
            onSelect(value);
        }
    };

    render() {
        const {
            minWidth,
            icon,
            children,
        } = this.props;

        return (
            <li
                className={suggestionStyles.suggestionItem}
                style={{minWidth: minWidth + 'px'}}
            >
                <button
                    className={suggestionStyles.suggestion}
                    onClick={this.handleClick}
                >
                    {icon &&
                        <Icon
                            className={suggestionStyles.icon}
                            name={icon}
                        />
                    }
                    {typeof children === 'string' &&
                        this.highlightMatchingTextPart(children)
                    }
                    {typeof children === 'function' &&
                        children(this.highlightMatchingTextPart)
                    }
                </button>
            </li>
        );
    }
}
