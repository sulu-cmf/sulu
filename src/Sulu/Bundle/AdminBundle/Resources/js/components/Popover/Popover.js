// @flow
import React, {Fragment} from 'react';
import {Portal} from 'react-portal';
import {observer} from 'mobx-react';
import {action, computed, observable} from 'mobx';
import type {ElementRef, Node} from 'react';
import {afterElementsRendered} from '../../services/DOM';
import Backdrop from '../Backdrop';
import type {PopoverDimensions} from './types';
import PopoverPositioner from './PopoverPositioner';
import popoverStyles from './popover.scss';

type Props = {
    open: boolean,
    children?: (
        setPopoverElementRef: (ref: ElementRef<*>) => void,
        style: Object,
        verticalPosition: string,
    ) => Node,
    onClose?: () => void,
    /** This element will be used to position the popover */
    anchorElement: ElementRef<*>,
    centerChildElement?: ElementRef<*>,
    horizontalOffset: number,
    verticalOffset: number,
    backdrop: boolean,
};

@observer
export default class Popover extends React.Component<Props> {
    static defaultProps = {
        open: false,
        horizontalOffset: 0,
        verticalOffset: 0,
        backdrop: true,
    };

    @observable popoverChildRef: ElementRef<'div'>;

    @observable popoverWidth: number;

    @observable popoverHeight: number;

    componentDidMount() {
        window.addEventListener('blur', this.close);
        window.addEventListener('resize', this.close);
    }

    componentWillUnmount() {
        window.removeEventListener('blur', this.close);
        window.removeEventListener('resize', this.close);
    }

    componentDidUpdate() {
        if (this.popoverChildRef) {
            this.updateDimensions();

            afterElementsRendered(() => {
                this.popoverChildRef.scrollTop = this.dimensions.scrollTop;
            });
        }
    }

    close = () => {
        if (this.props.onClose) {
            this.props.onClose();
        }
    };

    @computed get dimensions(): PopoverDimensions {
        const {
            anchorElement,
            verticalOffset,
            horizontalOffset,
            centerChildElement,
        } = this.props;
        const {
            top = 0,
            left = 0,
            width = 0,
            height = 0,
        } = anchorElement.getBoundingClientRect();
        const centerChildOffsetTop = (centerChildElement) ? centerChildElement.offsetTop : 0;
        const alignOnVerticalAnchorEdges = !centerChildElement;

        return PopoverPositioner.getCroppedDimensions(
            this.popoverWidth,
            this.popoverHeight,
            top,
            left,
            width,
            height,
            horizontalOffset,
            verticalOffset,
            centerChildOffsetTop,
            alignOnVerticalAnchorEdges
        );
    }

    updateDimensions = () => {
        if (!this.popoverChildRef) {
            return;
        }

        const {
            offsetWidth,
            offsetHeight,
            scrollWidth,
            scrollHeight,
            clientWidth,
            clientHeight,
        } = this.popoverChildRef;

        // calculating real size by considering borders, margins and paddings
        const outerWidth = scrollWidth + offsetWidth - clientWidth;
        const outerHeight = scrollHeight + offsetHeight - clientHeight;

        this.setPopoverSize(
            outerWidth,
            outerHeight
        );
    };

    @action setPopoverSize(width: number, height: number) {
        this.popoverWidth = width;
        this.popoverHeight = height;
    }

    handleBackdropClick = this.close;

    @action setPopoverChildRef = (popoverChildRef: ElementRef<*>) => {
        if (popoverChildRef) {
            this.popoverChildRef = popoverChildRef;
        }
    };

    render() {
        const {
            open,
            children,
            anchorElement,
            backdrop,
        } = this.props;

        if (!open || !anchorElement) {
            return null;
        }

        const dimensions = this.dimensions;
        const styles = {
            ...PopoverPositioner.dimensionsToStyle(dimensions),
            position: 'fixed',
            pointerEvents: 'auto',
        };

        const verticalPosition = (dimensions.top > anchorElement.getBoundingClientRect().top) ? 'bottom' : 'top';

        return (
            <Fragment>
                {backdrop &&
                    <Backdrop visible={false} open={true} onClick={this.handleBackdropClick} />
                }
                <Portal>
                    <div className={popoverStyles.container}>
                        {children &&
                            children(this.setPopoverChildRef, styles, verticalPosition)
                        }
                    </div>
                </Portal>
            </Fragment>
        );
    }
}
