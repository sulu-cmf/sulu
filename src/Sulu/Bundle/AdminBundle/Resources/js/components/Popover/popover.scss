/* to disable scrolling */
.container {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 110;
    pointer-events: none;
}
