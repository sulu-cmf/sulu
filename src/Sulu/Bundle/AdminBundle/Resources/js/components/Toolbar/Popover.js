// @flow
import React from 'react';
import type {ElementRef} from 'react';
import {action, observable} from 'mobx';
import {observer} from 'mobx-react';
import classNames from 'classnames';
import PopoverComponent from '../Popover';
import type {Popover as PopoverProps} from './types';
import Button from './Button';
import popoverStyles from './popover.scss';

@observer
class Popover extends React.Component<PopoverProps> {
    @observable open: boolean = false;

    static defaultProps = {
        showText: true,
    };

    @observable buttonRef: ?ElementRef<'button'>;

    @action setButtonRef = (ref: ?ElementRef<'button'>) => {
        if (ref) {
            this.buttonRef = ref;
        }
    };

    @action close = () => {
        this.open = false;
    };

    @action toggle = () => {
        this.open = !this.open;
    };

    componentDidUpdate() {
        const {disabled} = this.props;

        if (disabled) {
            this.close();
        }
    }

    handleButtonClick = () => {
        this.toggle();
    };

    handlePopoverClose = () => {
        this.close();
    };

    render() {
        const {
            children,
            className,
            icon,
            size,
            skin,
            label,
            disabled,
            loading,
            showText,
        } = this.props;
        const popoverClass = classNames(
            className,
            popoverStyles.popover,
            {
                [popoverStyles[size]]: size,
            }
        );

        return (
            <div className={popoverClass}>
                <Button
                    active={this.open}
                    buttonRef={this.setButtonRef}
                    disabled={disabled}
                    hasOptions={true}
                    icon={icon}
                    label={showText ? label : undefined}
                    loading={loading}
                    onClick={this.handleButtonClick}
                    size={size}
                    skin={skin}
                />
                <PopoverComponent
                    anchorElement={this.buttonRef}
                    onClose={this.handlePopoverClose}
                    open={this.open}
                >
                    {
                        (setPopoverElementRef, popoverStyle) => (
                            <div ref={setPopoverElementRef} style={popoverStyle}>
                                {children(this.close)}
                            </div>
                        )
                    }
                </PopoverComponent>
            </div>
        );
    }
}

export default Popover;
