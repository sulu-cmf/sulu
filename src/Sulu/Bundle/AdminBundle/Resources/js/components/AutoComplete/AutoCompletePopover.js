// @flow
import React from 'react';
import type {ElementRef} from 'react';
import Menu from '../Menu';
import Popover from '../Popover';
import Suggestion from './Suggestion';
import autoCompletePopoverStyles from './autoCompletePopover.scss';

type Props = {
    anchorElement: ElementRef<*>,
    onSelect: (suggestion: Object) => void,
    open: boolean,
    query: ?string,
    searchProperties: Array<string>,
    suggestions: Array<Object>,
};

export default class AutoCompletePopover extends React.Component<Props> {
    handlePopoverClose = () => {
        const {onSelect, suggestions} = this.props;
        if (suggestions.length > 0) {
            onSelect(suggestions[0]);
        }
    };

    render() {
        const {
            anchorElement,
            onSelect,
            open,
            query,
            searchProperties,
            suggestions,
        } = this.props;

        return (
            <Popover
                open={open}
                onClose={this.handlePopoverClose}
                anchorElement={anchorElement}
                verticalOffset={-2}
                horizontalOffset={5}
            >
                {
                    (setPopoverElementRef, popoverStyle) => (
                        <Menu
                            style={popoverStyle}
                            menuRef={setPopoverElementRef}
                        >
                            {suggestions.map((searchResult) => (
                                <Suggestion
                                    key={searchResult.id}
                                    onSelect={onSelect}
                                    query={query}
                                    value={searchResult}
                                >
                                    {(highlight) => searchProperties.map((field) => (
                                        <span className={autoCompletePopoverStyles.column} key={field}>
                                            {highlight(searchResult[field])}
                                        </span>
                                    ))}
                                </Suggestion>
                            ))}
                        </Menu>
                    )
                }
            </Popover>
        );
    }
}
