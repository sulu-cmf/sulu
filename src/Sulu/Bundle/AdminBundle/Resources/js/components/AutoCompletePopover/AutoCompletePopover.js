// @flow
import React from 'react';
import type {ElementRef} from 'react';
import Menu from '../Menu';
import Popover from '../Popover';
import Suggestion from './Suggestion';
import autoCompletePopoverStyles from './autoCompletePopover.scss';

type Props = {
    anchorElement: ElementRef<*>,
    minWidth: number,
    onSelect: (suggestion: Object) => void,
    open: boolean,
    query: ?string,
    searchProperties: Array<string>,
    suggestions: Array<Object>,
};

export default class AutoCompletePopover extends React.Component<Props> {
    static defaultProps = {
        minWidth: 0,
    };

    handlePopoverClose = () => {
        const {onSelect, suggestions} = this.props;
        if (suggestions.length > 0) {
            onSelect(suggestions[0]);
        }
    };

    render() {
        const {
            anchorElement,
            minWidth,
            onSelect,
            open,
            query,
            searchProperties,
            suggestions,
        } = this.props;

        return (
            <Popover
                open={open}
                onClose={this.handlePopoverClose}
                anchorElement={anchorElement}
                verticalOffset={-2}
                horizontalOffset={5}
            >
                {
                    (setPopoverElementRef, popoverStyle) => (
                        <Menu
                            style={popoverStyle}
                            menuRef={setPopoverElementRef}
                        >
                            {suggestions.map((searchResult) => (
                                <Suggestion
                                    key={searchResult.id}
                                    minWidth={minWidth}
                                    onSelect={onSelect}
                                    query={query}
                                    value={searchResult}
                                >
                                    {(highlight) => searchProperties.map((field) => (
                                        <span className={autoCompletePopoverStyles.column} key={field}>
                                            {highlight(searchResult[field])}
                                        </span>
                                    ))}
                                </Suggestion>
                            ))}
                        </Menu>
                    )
                }
            </Popover>
        );
    }
}
