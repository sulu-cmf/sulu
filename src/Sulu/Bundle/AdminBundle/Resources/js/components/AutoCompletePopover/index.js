// @flow
import AutoCompletePopover from './AutoCompletePopover';

export default AutoCompletePopover;
