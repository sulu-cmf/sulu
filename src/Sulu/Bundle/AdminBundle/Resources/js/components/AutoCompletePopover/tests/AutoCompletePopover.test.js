// @flow
import React from 'react';
import {mount, render, shallow} from 'enzyme';
import Mousetrap from 'mousetrap';
import AutoCompletePopover from '../AutoCompletePopover';
import Popover from '../../Popover';

jest.mock('../../Popover', () => ({children}) => children(jest.fn(), {}));

beforeEach(() => {
    Mousetrap.reset();
});

test('Popover should be hidden when open is set to false', () => {
    const autoCompletePopover = shallow(
        <AutoCompletePopover
            anchorElement={jest.fn()}
            onSelect={jest.fn()}
            open={false}
            query=""
            searchProperties={[]}
            suggestions={[]}
        />
    );

    expect(autoCompletePopover.find(Popover).prop('open')).toEqual(false);
});

test('Popover should be shown when open is set to true', () => {
    const autoCompletePopover = shallow(
        <AutoCompletePopover
            anchorElement={jest.fn()}
            onSelect={jest.fn()}
            open={true}
            query=""
            searchProperties={[]}
            suggestions={[]}
        />
    );

    expect(autoCompletePopover.find(Popover).prop('open')).toEqual(true);
});

test('Render with highlighted suggestions', () => {
    const suggestions = [
        {id: 1, name: 'Test 1'},
        {id: 2, name: 'Test 2'},
    ];

    expect(render(
        <AutoCompletePopover
            anchorElement={jest.fn()}
            onSelect={jest.fn()}
            open={true}
            query="Test"
            searchProperties={['name']}
            suggestions={suggestions}
        />
    )).toMatchSnapshot();
});

test('Call onSelect with first suggestion on close', () => {
    const suggestions = [
        {id: 1, name: 'Test 1'},
        {id: 2, name: 'Test 2'},
    ];

    const selectSpy = jest.fn();
    const autoCompletePopover = shallow(
        <AutoCompletePopover
            anchorElement={jest.fn()}
            onSelect={selectSpy}
            open={true}
            query="Test"
            searchProperties={['name']}
            suggestions={suggestions}
        />
    );

    autoCompletePopover.find(Popover).prop('onClose')();
    expect(selectSpy).toBeCalledWith(suggestions[0]);
});

test('Call onSelect with clicked suggestion', () => {
    const suggestions = [
        {id: 1, name: 'Test 1'},
        {id: 2, name: 'Test 2'},
    ];

    const selectSpy = jest.fn();
    const autoCompletePopover = mount(
        <AutoCompletePopover
            anchorElement={jest.fn()}
            onSelect={selectSpy}
            open={true}
            query="Test"
            searchProperties={['name']}
            suggestions={suggestions}
        />
    );

    expect(autoCompletePopover.find('Suggestion').at(1).prop('value')).toBe(suggestions[1]);
    autoCompletePopover.find('Suggestion').at(1).prop('onSelect')(suggestions[1]);
    expect(selectSpy).toBeCalledWith(suggestions[1]);
});

test('Pressing down should select next item', () => {
    const suggestions = [
        {id: 1, name: 'Test 1'},
        {id: 2, name: 'Test 2'},
    ];

    const autoCompletePopover = mount(
        <AutoCompletePopover
            anchorElement={jest.fn()}
            onSelect={jest.fn()}
            open={false}
            query="Test"
            searchProperties={['name']}
            suggestions={suggestions}
        />
    );

    const suggestionElement1 = {focus: jest.fn()};
    const suggestionElement2 = {focus: jest.fn()};

    autoCompletePopover.instance().suggestionsRef = {
        getElementsByTagName: jest.fn().mockReturnValue([
            suggestionElement1,
            suggestionElement2,
        ]),
    };
    autoCompletePopover.setProps({open: true});
    autoCompletePopover.update();

    Mousetrap.trigger('down');
    expect(suggestionElement1.focus).toBeCalledWith();
    expect(suggestionElement2.focus).not.toBeCalledWith();
});

test('Pressing up should select previous item', () => {
    const suggestions = [
        {id: 1, name: 'Test 1'},
        {id: 2, name: 'Test 2'},
    ];

    const autoCompletePopover = mount(
        <AutoCompletePopover
            anchorElement={jest.fn()}
            onSelect={jest.fn()}
            open={false}
            query="Test"
            searchProperties={['name']}
            suggestions={suggestions}
        />
    );

    const suggestionElement1 = {focus: jest.fn()};
    const suggestionElement2 = {focus: jest.fn()};

    // $FlowFixMe
    Object.defineProperty(document, 'activeElement', {
        value: suggestionElement2,
    });

    autoCompletePopover.instance().suggestionsRef = {
        getElementsByTagName: jest.fn().mockReturnValue([
            suggestionElement1,
            suggestionElement2,
        ]),
    };
    autoCompletePopover.setProps({open: true});
    autoCompletePopover.update();

    Mousetrap.trigger('up');
    expect(suggestionElement1.focus).toBeCalledWith();
    expect(suggestionElement2.focus).not.toBeCalledWith();
});
