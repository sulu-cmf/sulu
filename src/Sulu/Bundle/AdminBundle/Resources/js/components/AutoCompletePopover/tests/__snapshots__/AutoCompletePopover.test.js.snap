// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render with highlighted suggestions 1`] = `
<ul
  class="menu"
>
  <li
    class="suggestionItem"
    style="min-width:0px"
  >
    <button
      class="suggestion"
    >
      <span
        class="column"
      >
        <span>
          <strong>
            Test
          </strong>
           1
        </span>
      </span>
    </button>
  </li>
  <li
    class="suggestionItem"
    style="min-width:0px"
  >
    <button
      class="suggestion"
    >
      <span
        class="column"
      >
        <span>
          <strong>
            Test
          </strong>
           2
        </span>
      </span>
    </button>
  </li>
</ul>
`;
